### PR TITLE
fix(cli): Create child node_modules for conflicting dependency versions, respect aliases in package.json

### LINKED ISSUE #24419: npm workspaces: understand when to create child node_modules directories
In the upcoming npm workspaces implementation, it's not smart enough to know when to create child node_modules directories. It needs to do this when there's conflicts between workspace packages.

### PR DESCRIPTION
**Pull request by nathanwhit**
Fixes #24419.

### CODE REVIEW

File: cli/npm/managed/resolvers/local.rs [hunk 1]
```diff
@@ -609,16 +609,79 @@ async fn sync_resolution_with_fs(
...
+    match found_names.entry(&id.nv.name) {
+      Entry::Occupied(_) => {
+        continue; // skip, already handled
+      }
+      Entry::Vacant(entry) => {
+        entry.insert(&id.nv);
+      }
     }
     let package = snapshot.package_from_id(id).unwrap();
     let target_folder_name =
```
Comment (nathanwhit) [inline cli/npm/managed/resolvers/local.rs:687]: I was hoping to be able to just remove this section and just use the remote + workspace reqs, but from what I could tell they only include deps from package.json, not ones from npm specifiers.

File: cli/npm/managed/resolvers/local.rs [hunk 2]
```diff
@@ -609,16 +609,79 @@ async fn sync_resolution_with_fs(
...
-  // node_modules/.deno/<package_id>/node_modules/<package_name>
-  let mut found_names = HashSet::new();
-  let mut ids = snapshot.top_level_packages().collect::<Vec<_>>();
+  let mut found_names: HashMap<&String, &PackageNv> = HashMap::new();
+
+  // 4. Create symlinks for package json dependencies
+  {
+    for remote in pkg_json_deps_provider.remote_pkgs() {
+      let remote_id = snapshot
+        .resolve_best_package_id(&remote.req.name, &remote.req.version_req)
```
Comment (nathanwhit) [inline cli/npm/managed/resolvers/local.rs:618]: Is there a better way to resolve the version req to a package id?

File: cli/npm/managed/resolvers/local.rs [hunk 3]
```diff
@@ -609,16 +609,79 @@ async fn sync_resolution_with_fs(
...
+          .join(&target_folder_name)
+          .join("node_modules"),
+        &remote_pkg.id.nv.name,
+      );
+      if install_in_child {
+        // symlink the dep into the package's child node_modules folder
+        let dest_path =
+          remote.base_dir.join("node_modules").join(&remote.alias);
+
+        symlink_package_dir(&local_registry_package_path, &dest_path)?;
```
Comment (nathanwhit) [inline cli/npm/managed/resolvers/local.rs:653]: Should probably add to the setup cache

File: cli/npm/managed/resolvers/local.rs (hunk 2 above)
Comment (dsherret) [reply cli/npm/managed/resolvers/local.rs:618]: This is the way to do it. I'm not sure why the tests are panicking on this line though.

**Review by dsherret — COMMENTED** [review]
Very nice! This looks good to me, but I'm not sure why the CI is failing because of it.

File: cli/npm/managed/resolvers/local.rs (hunk 2 above)
Comment (nathanwhit) [reply cli/npm/managed/resolvers/local.rs:618]: Ah if you have
```
deno cache npm:@types/node
```

but your package.json contains a dep on `@denotest/add`, then the req will still be there, but `@denotest/add` won't be in the snapshot

Comment (dsherret) [reply cli/npm/managed/resolvers/local.rs:618]: Ah, yeah if it doesn't match a package in the package.json then it won't auto-install the package.json dependencies.

**Review by dsherret — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**